mpegts/mpegts_demuxer.cpp

```diff
@@ -97,85 +97,87 @@ uint8_t MpegTsDemuxer::decode(SimpleBuffer &rIn) {
+                EsFrame* lEsFrame = mEsFrames[lTsHeader.mPid].get();
+                if (lTsHeader.mPayloadUnitStartIndicator == 0x01) {
+                    // TODO: Transport streams with no PES packet size set on video streams will not output the very
+                    // last frame, since there will be no final packet with mPayloadUnitStartIndicator set. Add a flush
+                    // function to the demuxer API that the user can call once it has ran out of input data.
 
-                    if (mEsFrames[lTsHeader.mPid]->mCompleted) {
-                        mEsFrames[lTsHeader.mPid]->reset();
-                    } else if (mEsFrames[lTsHeader.mPid]->mData->size() && !mEsFrames[lTsHeader.mPid]->mCompleted) {
-                        //Its a broken frame deliver that as broken
+                    if (!lEsFrame->mData->empty()) {
+                        // We have an old packet with data, flush and reset it before starting a new one
                         if (esOutCallback) {
-                            EsFrame *lEsFrame = mEsFrames[lTsHeader.mPid].get();
-                            lEsFrame -> mBroken = true;
-                            lEsFrame -> mPid = lTsHeader.mPid;
+                            if (lEsFrame->mExpectedPesPacketLength == 0) {
+                                lEsFrame->mCompleted = true;
+                            } else {
+                                // Expected length is set, but we didn't get enough data, deliver what we have as broken
+                                lEsFrame->mBroken = true;
+                            }
                             esOutCallback(lEsFrame);
                         }
-
-                        mEsFrames[lTsHeader.mPid]->reset();
+                        lEsFrame->reset();
                     }
 
+                    // Potential old package has been flushed, set random access indicator and start new package
+                    lEsFrame->mRandomAccess = lRandomAccessIndicator;
+
+                    PESHeader lPesHeader;
                     lPesHeader.decode(rIn);
-                    mEsFrames[lTsHeader.mPid]->mStreamId = lPesHeader.mStreamId;
-                    mEsFrames[lTsHeader.mPid]->mExpectedPesPacketLength = lPesHeader.mPesPacketLength;
+                    lEsFrame->mStreamId = lPesHeader.mStreamId;
+                    lEsFrame->mExpectedPesPacketLength = lPesHeader.mPesPacketLength;
                     if (lPesHeader.mPtsDtsFlags == 0x02) {
-                        mEsFrames[lTsHeader.mPid]->mPts = mEsFrames[lTsHeader.mPid]->mDts = readPts(rIn);
+                        lEsFrame->mPts = lEsFrame->mDts = readPts(rIn);
                     } else if (lPesHeader.mPtsDtsFlags == 0x03) {
-                        mEsFrames[lTsHeader.mPid]->mPts = readPts(rIn);
-                        mEsFrames[lTsHeader.mPid]->mDts = readPts(rIn);
+                        lEsFrame->mPts = readPts(rIn);
+                        lEsFrame->mDts = readPts(rIn);

```

Comment:
Typo? Should it be `readDts`?

Reply:
Hah, nice catch, seems none has demuxed streams with B-frames and/or looked at DTS:es.
We currently just use the PTS in the mediasource. Fixing

Reply:
..oh just saw that it is just the function name, reading a pts/dts would be the same code .. hence this was actually correct, but ofc a bit confusing

Reply:
Ah, I saw that now too. Maybe change the name of this?